File: src/rtweekendpkg/hittable.nim
import ray
import vec3

import std/options

type HitRecord* = object
  p*: Point3
  normal*: Vec3
  t*: float
  frontFace*: bool

proc newHitRecord*(p: Point3, t: float): HitRecord =
  HitRecord(
    p: p,
    normal: newVec3(),
    t: t,
    frontFace: false
  )

proc setFaceNormal*(self: var HitRecord, r: Ray, outwardNormal: Vec3) =
  self.frontFace = r.direction.dot(outwardNormal) < 0
  self.normal =
    if self.frontFace:
      outwardNormal
    else:
      -outwardNormal

type Hittable* = ref object of RootObj

method hit*(
  h: Hittable,
  r: Ray,
  tMin: float,
  tMax: float,
): Option[HitRecord] {.base.} =
  raiseAssert "hit() not implemented"

File: src/hittable.rs
use crate::{
    ray::Ray,
    vec3::{dot, Point3, Vec3},
};

pub(crate) struct HitRecord {
    pub(crate) p: Point3,
    pub(crate) normal: Vec3,
    pub(crate) t: f64,
    pub(crate) front_face: bool,
}

impl HitRecord {
    pub(crate) fn new(p: Point3, t: f64, r: &Ray, outward_normal: Vec3) -> Self {
        let front_face = dot(r.direction(), &outward_normal) < 0.0;
        let normal = if front_face {
            outward_normal
        } else {
            -outward_normal
        };
        Self {
            p,
            normal,
            t,
            front_face,
        }
    }
}

pub(crate) trait Hittable {
    fn hit(&self, r: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord>;
}
